File: packages/sdk-python/src/agentmark_sdk/serialize.py
"""Serialization utilities for observed function IO capture."""

from __future__ import annotations

import dataclasses
import json
from typing import Any

MAX_SERIALIZE_LENGTH = 1_000_000


def serialize_value(value: Any, max_length: int = MAX_SERIALIZE_LENGTH) -> str:
    """Serialize a value to a JSON string for span attributes.

    Serialization chain:
      1. Pydantic model → .model_dump() → JSON
      2. Dataclass → dataclasses.asdict() → JSON
      3. Dict/list/primitive → JSON directly
      4. Fallback → str(obj)

    Truncated to max_length characters.
    """
    try:
        if hasattr(value, "model_dump"):
            serialized = json.dumps(value.model_dump(), default=str)
        elif dataclasses.is_dataclass(value) and not isinstance(value, type):
            serialized = json.dumps(dataclasses.asdict(value), default=str)
        else:
            serialized = json.dumps(value, default=str)
    except (TypeError, ValueError, OverflowError):
        serialized = str(value)

    return serialized[:max_length]

File: packages/sdk/src/trace/serialize.ts
/**
 * Serialization utilities for observed function IO capture.
 */

export const MAX_SERIALIZE_LENGTH = 1_000_000;

/**
 * Serialize a value to a JSON string for span attributes.
 *
 * Serialization chain:
 *   1. Object with toJSON() → JSON
 *   2. Plain object/array/primitive → JSON
 *   3. Fallback → String(obj)
 *
 * Truncated to maxLength characters.
 */
export function serializeValue(
  value: unknown,
  maxLength: number = MAX_SERIALIZE_LENGTH
): string {
  let serialized: string;
  try {
    serialized = JSON.stringify(value, (_key, val) => {
      // Handle BigInt
      if (typeof val === "bigint") {
        return val.toString();
      }
      return val;
    });
  } catch {
    serialized = String(value);
  }
  return serialized.slice(0, maxLength);
}
